### PR TITLE
feat: spawn enemies

### PR DESCRIPTION
**Pull request by Tboefijn**
## Description

I have added the spawning of enemies and a more intuitive method of enemy vision. The enemy will detect you if you are in a certain path distance. When the player is detected, the enemy will walk to the player. If the player is fast enough and escapes outside the vision of the enemy, the enemy will walk towards the last known location of the player.

## Testability

Open and play the GridSystem scene.

## Checklist
- [X] Set the proper pull request name
- [X] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

**Review by joachimdekker — CHANGES_REQUESTED** [review]
I have left a few comments. I have to still test it as well.

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs [hunk 1]
```diff
@@ -21,10 +21,16 @@ public class MeleeEnemy : DummyEnemy
     [SerializeField]
     private float _swingLength = 4f;
 
+    [SerializeField]
+    private int _visionRange = 20;
```
[the remark this answers is not in the document]
Comment (Tboefijn) [reply aplib.net-demo/Assets/Scripts/Entities/Enemy/MeleeEnemy.cs]: Had already deleted the using of it but not the variable itself.

**Review by joachimdekker — DISMISSED** [review]

File: aplib.net-demo/Assets/Scripts/WFC/GameObjectPlacer.cs [hunk 2]
```diff
@@ -105,22 +94,10 @@ public void Initialize()
...
-        public (int x, int z) GetCellCoordinates(Vector3 position) =>
-            ((int)(position.x / _tileSizeX), (int)(position.z / _tileSizeZ));
 
-        /// <summary>
-        /// Calculates the center of a cell's floor in world coordinates.
-        /// </summary>
-        /// <param name="cell">The cell to calculate its center of.</param>
-        /// <returns>The real-world coordinates of the center of the cell's floor.</returns>
-        public Vector3 CenterOfCell(Cell cell) => new(cell.X * _tileSizeX, 0, cell.Z * _tileSizeZ);
+            if (!TryGetComponent(out _spawningExtensions))
```
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/WFC/GameObjectPlacer.cs:98]: Dit is niet nodig met RequireComponent

File: aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs [hunk 3]
```diff
@@ -75,6 +88,14 @@ public void Awake()
             _gameObjectPlacer = GetComponent<GameObjectPlacer>();
             _gameObjectPlacer.Initialize();
 
+            if (!TryGetComponent(out _spawningExtensions))
+                throw new UnityException("SpawningExtensions not found.");
+
+            if (!TryGetComponent(out _enemySpawner))
+                throw new UnityException("EnemySpawner not found.");
```
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/WFC/GridPlacer.cs:95]: Niet nodig met RequireComponent

File: aplib.net-demo/Assets/Scripts/WFC/SpawnAbleEnemies.cs
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/WFC/SpawnAbleEnemies.cs:1]: Kan naar scriptable objects folder toch?

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/AbstractEnemy.cs [hunk 4]
```diff
@@ -10,9 +10,17 @@ public abstract class AbstractEnemy : MonoBehaviour
...
-    [SerializeField] protected string _targetTag = "Player";
+    [SerializeField]
+    protected int _damagePoints = 25;
+
+    [SerializeField]
+    protected string _targetTag = "Player";
+
+    [SerializeField]
+    private int _defaultVisionRange = 25;
+
```
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/AbstractEnemy.cs:21]: maak protected zodat ranged enemy kan hergebruiken

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs:1]: De vision range in deze klasse kan vervangen worden door de defaultVisionRange van de abstract enemy toch?

(en de TagToFind setting in de constructor kan vgm ook weg)

File: aplib.net-demo/Assets/Scripts/MainMenu/CanvasManager.cs [hunk 5]
```diff
@@ -51,19 +51,19 @@ public class CanvasManager : MonoBehaviour
...
     /// </summary>
-    private bool _isOnGameSettings;
+    private bool _preventSettingsToggle;
 
     /// <summary>
-    /// To ensure the menu settings and menu UI aren't on at the same time.
+    /// To ensure the game settings and menu UI aren't on at the same time.
     /// </summary>
-    private bool _isOnMenuSettings;
+    public bool IsOnGameSettings { get; private set; }
```
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/MainMenu/CanvasManager.cs:61]: waarom public gemaakt? Had je die nodig?

**Review by JoenvandeVorle — CHANGES_REQUESTED** [review]

File: aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs [hunk 6]
```diff
@@ -44,6 +46,8 @@ protected override void Start()
     /// </summary>
     protected override void Update()
     {
+        Debug.DrawRay(transform.position, Vector3.up * 20, Color.green);
+
         // If the target is not within vision range, do nothing.
         if (!_pathFind.GoalWithinRange(_visionRange))
         {
```
Comment (JoenvandeVorle) [inline aplib.net-demo/Assets/Scripts/Entities/Enemy/RangedEnemy.cs:53]: hele update kan anders met nieuwe pathfind, movecloser coroutine kan weg
```suggestion
    protected override void Update()
    {
        Debug.DrawRay(transform.position, Vector3.up * 20, Color.green);

        // If the target is not directly visible but within vision range, move closer to the target.
        if (!_rangedWeapon.EnemiesInLineOfSight() && _pathFind.GoalWithinRange(_defaultVisionRange))
        {
            _pathFind.SetStoppingDistance(1f);
        }

        if (_attackTimer.IsFinished() && _rangedWeapon.EnemiesInLineOfSight())
        {
            _pathFind.SetStoppingDistance(_attackRange - 1f);
            _rangedWeapon.UseWeapon();
            _attackTimer.Reset();
        }

        base.Update();
    }
```

File: aplib.net-demo/Assets/Scripts/WFC/SpawnAbleEnemies.cs
Comment (Tboefijn) [reply aplib.net-demo/Assets/Scripts/WFC/SpawnAbleEnemies.cs:1]: Daar staan de assets zelf en niet de scripts.

File: aplib.net-demo/Assets/Scripts/MainMenu/CanvasManager.cs (hunk 5 above)
Comment (Tboefijn) [reply aplib.net-demo/Assets/Scripts/MainMenu/CanvasManager.cs:61]: Ja in GridPlacerEditor.cs

**Review by JoenvandeVorle — CHANGES_REQUESTED** [review]
nog 1 kleine nit

**Review by JoenvandeVorle — APPROVED** [review]
awimbo!

**Review by joachimdekker — APPROVED** [review]

**Comment by JoenvandeVorle** [comment]
WAIT

geef de enemies drag, dan bouncen ze niet meer zo

**Review by JoenvandeVorle — CHANGES_REQUESTED** [review]
zie comment

**Review by joachimdekker — APPROVED** [review]

**Review by JoenvandeVorle — DISMISSED** [review]

**Review by SilasPeters — APPROVED** [review]

**Review by QuakeEye — APPROVED** [review]

**Review by SSelinn — DISMISSED** [review]

**Review by QuakeEye — APPROVED** [review]

**Review by SSelinn — APPROVED** [review]

**Review by SilasPeters — APPROVED** [review]

**Review by SSelinn — APPROVED** [review]
I

**Review by QuakeEye — APPROVED** [review]
image

**Pull request merged**